server/worker/tasks.py

```diff
@@ -0,0 +1,112 @@
+import uuid
+import traceback
+import json
+import logging
+from datetime import datetime
+from typing import Optional
+from sqlalchemy import update
+
+from ..database import db_session
+from ..models import *  # pylint: disable=wildcard-import
+from ..api import rounds
+from ..util.isoformat import isoformat
+from ..util.jsonschema import JSONDict
+from ..config import FLASK_ENV
+
+
+logger = logging.getLogger("arlo.worker")
+
+
+class UserError(Exception):
+    pass
+
+
+def create_background_task(task_name: TaskName, payload: JSONDict) -> BackgroundTask:
+    task = BackgroundTask(id=str(uuid.uuid4()), task_name=task_name, payload=payload)
+    db_session.add(task)
+    # In testing, we run the task immediately, since we'll only be doing small
+    # tasks and we want to make sure they run in a thread-safe way (i.e. we
+    # don't want tests to interfere with each other's background tasks when
+    # running concurrently).
+    if FLASK_ENV == "test":
+        db_session.flush()
+        run_task(task)
+    return task
+
+
+def serialize_background_task(task: Optional[BackgroundTask]) -> Optional[JSONDict]:
+    if task is None:
+        return None
+
+    if task.error:
+        status = ProcessingStatus.ERRORED
+    elif task.completed_at:
+        status = ProcessingStatus.PROCESSED
+    elif task.started_at:
+        status = ProcessingStatus.PROCESSING
+    else:
+        status = ProcessingStatus.READY_TO_PROCESS
+
+    return {
+        "status": status,
+        "startedAt": isoformat(task.started_at),
+        "completedAt": isoformat(task.completed_at),
+        "error": task.error,
+    }
+
+
+def run_task(task: BackgroundTask) -> bool:
+    print("Run", task.__dict__)
+    if task.started_at:
+        return False
+
+    # Claim this task by updating the `started_at` timestamp in such a way that
+    # it must not have been set before.
+    result = db_session.execute(
+        update(BackgroundTask.__table__)  # pylint: disable=no-member
+        .where(BackgroundTask.id == task.id)
+        .where(BackgroundTask.started_at.is_(None))
+        .values(started_at=datetime.now(timezone.utc))
+    )
+    if result.rowcount == 0:
```

Comment:
Unless this is its own transaction in this statement that commits right away, this will just block other threads during the whole process of running the task until the transaction commits. Is that what `db_session.execute` does?

If you're planning for multiple workers, we may also want to plan for cleanup on backend job crash that would otherwise leave tasks uncompleted.

Reply:
I actually just copied this over from the code that @eventualbuddha wrote for processing files, so I didn't take the time to think deeply about it.

I did a little research:
- db_session.execute is just a way to run a query without using the ORM layer, it doesn't do anything else special
- This statement executes in the current transaction, not in its own transaction
- Since the default transaction isolation level in Postgres is "read committed", I think that if there were multiple threads trying to execute this statement, you are correct that they would get blocked until the first thread to execute this statement commits the transaction (according to https://www.postgresql.org/docs/9.5/transaction-iso.html)

I'm not totally sure what you mean about planning for cleanup beyond what we already do (which is to rollback the nested transaction and then commit the outer transaction, recording the error).

That being said, maybe we don't need any of this since we're running in a single-worker context?

Reply:
right, if we continue to assume single-worker context (and if so we should make the explicitly clear in the code), we can simplify this.